File: src/Koinon.Domain/Entities/FollowUp.cs
using Koinon.Domain.Enums;

namespace Koinon.Domain.Entities;

/// <summary>
/// Represents a follow-up task for a person, typically a first-time visitor or new attendee.
/// Tracks the status and assignment of follow-up contact attempts.
/// </summary>
public class FollowUp : Entity
{
    /// <summary>
    /// Foreign key to the person who needs follow-up.
    /// </summary>
    public int PersonId { get; set; }

    /// <summary>
    /// Navigation property to the person who needs follow-up.
    /// </summary>
    public virtual Person? Person { get; set; }

    /// <summary>
    /// Optional foreign key to the attendance record that triggered this follow-up.
    /// </summary>
    public int? AttendanceId { get; set; }

    /// <summary>
    /// Navigation property to the attendance record.
    /// </summary>
    public virtual Attendance? Attendance { get; set; }

    /// <summary>
    /// Current status of the follow-up task.
    /// </summary>
    public FollowUpStatus Status { get; set; } = FollowUpStatus.Pending;

    /// <summary>
    /// Notes about the follow-up interaction, attempts, or outcomes.
    /// </summary>
    public string? Notes { get; set; }

    /// <summary>
    /// Optional foreign key to the person assigned to complete this follow-up.
    /// </summary>
    public int? AssignedToPersonId { get; set; }

    /// <summary>
    /// Navigation property to the person assigned to this follow-up task.
    /// </summary>
    public virtual Person? AssignedToPerson { get; set; }

    /// <summary>
    /// Date and time when contact was made with the person.
    /// </summary>
    public DateTime? ContactedDateTime { get; set; }

    /// <summary>
    /// Date and time when the follow-up was marked as complete.
    /// </summary>
    public DateTime? CompletedDateTime { get; set; }
}

File: src/web/src/types/followup.ts
/**
 * Follow-up Types
 *
 * TypeScript definitions for the follow-up management system.
 */

import type { IdKey, DateTime } from '@/services/api/types';

/**
 * Status of a follow-up record
 */
export enum FollowUpStatus {
  /** Initial state - not yet contacted */
  Pending = 0,
  /** Contact has been attempted or made */
  Contacted = 1,
  /** Multiple contact attempts with no response */
  NoResponse = 2,
  /** Successfully connected with the person */
  Connected = 3,
  /** Person declined further follow-up */
  Declined = 4,
}

/**
 * Represents a follow-up task for a person
 */
export interface FollowUpDto {
  /** Unique identifier for this follow-up record */
  idKey: IdKey;
  /** IdKey of the person being followed up with */
  personIdKey: IdKey;
  /** Name of the person being followed up with */
  personName: string;
  /** IdKey of the attendance record that triggered this follow-up (if applicable) */
  attendanceIdKey?: IdKey;
  /** Current status of the follow-up */
  status: FollowUpStatus;
  /** Notes about the follow-up */
  notes?: string;
  /** IdKey of the person assigned to this follow-up */
  assignedToIdKey?: IdKey;
  /** Name of the person assigned to this follow-up */
  assignedToName?: string;
  /** When contact was made */
  contactedDateTime?: DateTime;
  /** When the follow-up was completed */
  completedDateTime?: DateTime;
  /** When this follow-up record was created */
  createdDateTime: DateTime;
}

/**
 * Request to update the status of a follow-up record
 */
export interface UpdateFollowUpStatusRequest {
  /** New status for the follow-up */
  status: FollowUpStatus;
  /** Additional notes about the status update */
  notes?: string;
}

/**
 * Request to assign a follow-up to a person
 */
export interface AssignFollowUpRequest {
  /** IdKey of the person to assign this follow-up to */
  assignedToIdKey: IdKey;
}
